doc/man/doveadm-stats.1.in
.\" Copyright (c) 2010-2018 Dovecot authors, see the included COPYING file
.TH DOVEADM\-DUMP 1 "2016-01-12" "Dovecot v2.3" "Dovecot"
.SH NAME
doveadm\-stats \- Inspect or reset stats
.\"------------------------------------------------------------------------
.SH SYNOPSIS
.BR doveadm\ stats\ dump " [ " \-s\ <stats\ socket\ path> " ] " <type> " [ " <filter> " ]"
.PP
.BR doveadm\ stats\ top " [ " \-s\ <stats\ socket\ path> " ] " "[ " -b " ]" " [ " sort field> " ]"
.PP
.BR doveadm\ stats\ reset " [ " \-s\ <stats\ socket\ path> " ] "
.PP
.BR doveadm\ stats\ add " [ " \-\-description\ <string> " ] [ " \-\-exporter\ <name> " [ " \-\-exporter-include\ <field> " ]] [ " \-\-fields\ <fields> " ] [ " \-\-group_by\ <fields> " ] " <name>\ <filter>
.PP
.BR doveadm\ stats\ remove\ <name>
.\"------------------------------------------------------------------------
.SH DESCRIPTION
.B doveadm stats dump
is used to output statistics
.PP
.B doveadm stats top
is used to monitor statistics
.PP
.B doveadm stats reset
is used to reset statistics
.PP
.B doveadm stats add
is used to add metrics to statistics
.PP
.B doveadm stats remove
is used to remove metrics from statistics
.\"------------------------------------------------------------------------
.SH OPTIONS
Command specific
.IR options :
.TP
.BI \-s socketpath
Sets stats socket path
.TP
.B \-b
Show disk input/output bytes  
.\"------------------------------------------------------------------------
.SH ARGUMENTS
.BR dump
accepts following types: command, session, user, domain, ip and global. 
.PP
Filter can be
.TP
.B user=<wildcard>
Match given user.
.TP
.B domain=<wildcard>
Match given DNS domain name
.TP
.B session=<str>
Match session identifier
.TP
.B ip=<ip>"["/<mask>"]"
Match local or remote IP
.TP
.B since=<timestamp>
Match session start time
.TP
.B connected
Show only connected sessions
.PP
.BR top
accepts any valid field name to sort along with.
.\"------------------------------------------------------------------------
@INCLUDE:reporting-bugs@
.\"------------------------------------------------------------------------
.SH SEE ALSO
.BR doveadm (1)


NEWS
v2.3.17.1 2021-12-07  Aki Tuomi <aki.tuomi@open-xchange.com>

	- dsync: Add back accidentically removed parameters.
	- lib-ssl-iostream: Fix assert-crash when OpenSSL returned syscall error
	  without errno.
	- master: Dovecot failed to start if ssl_ca was too large.

v2.3.17 2021-10-28  Aki Tuomi <aki.tuomi@open-xchange.com>

	* Dovecot now logs a warning if time seems to jump forward at least
	  100 milliseconds.
	* dict: Lines logged by the dict process now contain the dict name as
	  the prefix.
	* lib-index: mail_cache_fields, mail_always_cache_fields and
	  mail_never_cache_fields now verifies that the listed header names are
	  valid. Especially the UTF8 "–" character has sometimes been wrongly
	  used instead of the ASCII "-".
	+ *-login: Added login_proxy_rawlog_dir setting to capture
	  rawlogs between proxy and backend.
	+ dict: The server process now keeps the last 10 idle dict backends
	  cached for maximum of 30 seconds. Practically this acts as a
	  connection pool for dict-redis and dict-ldap. Note that this doesn't
	  affect dict-sql, because it already had its own internal cache.
	+ doveadm: New stats add/remove commands added to support changing the
	  metrics configuration on runtime.
	+ lazy_expunge: Added lazy_expunge_exclude settings to disable
	  lazy_expunge for specific folders. \Special-use flags can be used as
	  folder names.
	+ lib-lua: Added a new helper function dovecot.restrict_global_variables()
	  to disable or enable defining new global variables.
	- LAYOUT=index List index rebuild was missing.
	- LAYOUT=index: Duplicate GUIDs were not detected.
	- acl: When using acl_ignore_namespace Dovecot attempted to access or
	  create dovecot-acl-list even when the namespace should have been
	  ignored. For virtual namespaces this could have yielded errors about
	  "Read-only file system" or "Permission denied".
	- auth: Setting the "master" passdb field to empty value would
	  cause proxying to fail with an authentication error.
	  Now an empty "master" field is ignored.
	- doveadm-server: Duplicate error lines were sent for failed commands.
	  This didn't normally cause visible problems, except when using
	  wildcards in usernames or -A parameter to go through multiple users.
	- doveadm-server: Logs written by doveadm-server were often missing log
	  prefixes, especially mail_log_prefix for mail commands. Logs sent to
	  doveadm TCP client were also missing log prefixes.
	- doveadm: v2.3 regression: batch command always crashes.
	- doveadm: v2.3.11 regression: Commands failed if ssl_cert or
	  ssl_key files weren't readable by the user running doveadm, even
	  though doveadm didn't actually use these settings
	- imap-hibernate: Process may crash at deinit:
	  Panic: file ioloop.c: line 928 (io_loop_destroy): assertion failed:
	  (ioloop->cur_ctx == NULL).
	- imap: Using imap_fetch_failure=no-after can cause assert-crash
	  with some IMAP commands if reading the mail fails (e.g. wrong cached
	  mail size). Fixes:
	  Panic: file index-mail-headers.c: line 198 (index_mail_parse_header_init):
	  assertion failed: (!mail->data.header_parser_initialized)
	- imap: v2.3.10 regression: When using INDEXPVT to enable private
	  \Seen flags (for shared or public namespaces) the STORE command did
	  not send untagged replies for the \Seen flag changes.
	- imap: v2.3.15 regression: If PREVIEW/SNIPPET is not the final FETCH
	  option in the command, the IMAP FETCH response is broken.
	- imap: v2.3.15 regression: MOVE command leaks mailbox if it can't be
	  opened and crashes at deinit:
	  Panic: file mail-user.c: line 229 (mail_user_deinit): assertion failed:
	  ((*user)->refcount == 1).
	- imapc: Copying nonexistent mail via imapc could have crashed. Fixes:
	  Panic: file mail-storage.c: line 2385 (mailbox_transaction_commit_get_changes):
	  assertion failed: (ret < 0 || seq_range_count(&changes_r->saved_uids) == save_count ||
	  array_count(&changes_r->saved_uids) == 0).
	- indexer: v2.3.15 regression: Process crashes if indexer-client
	  disconnects while it's waiting for command reply. This happened for
	  example if IMAP SEARCH triggered long fts indexing and the IMAP
	  client disconnected while waiting for the reply.
	- indexer: v2.3.15 regression: Process may have crashed in some situations.
	- indexer: v2.3.15 regression: indexer-worker processes may not have
	  reached the process_limit in some situations, possibly even using just
	  one indexer-worker process even though there were many indexing
	  requests queued.
	- lib-compression: Reading lz4 compressed mdbox mails may crash. Fixes:
	  Panic: file istream.c: line 345 (i_stream_read_memarea):
	  assertion failed: (!stream->blocking).
	- lib-compression: bench-compress crashes due to xz being read-only.
	- lib-lua: Fix linking libdict_lua for non-GNU linkers when Lua support
	  is disabled.
	- lib-mail: There was no limit on how large an email header name could be.
	  Processable header names are now limited to 1000 bytes.
	- lib-oauth2: Dovecot disallowed JWT tokens if their validity time was
	  older than token creation time (nbf < iat).
	- lib-storage: Reduce memory footprint of certain storage operations.
	- lib-storage: When listing mailboxes with storage name escape
	  characters (^ or .) as part of the mailbox name, the listing could
	  show corrupted mailbox names. Due to an issue in handling escaped
	  parent folders, the listing of other mailbox names would become
	  corrupted by prepending parts of the previously listed mailboxes
	  parent folder as prefix to the actual mailbox names. The corruption
	  can occur when using LAYOUT=INDEX and maildir or obox, or when using
	  the listescape plugin.
	- mail-crypt: Fix "-O" argument for "doveadm mailbox cryptokey password"
	  command to be a boolean, and not expect a string.
	- submission-login: Add support for not authenticating to next hop in
	  submission proxying.
	- submission-login: EHLO was not sent again after XCLIENT when doing

src/doveadm/doveadm-stats.c
/* Copyright (c) 2017-2018 Dovecot authors, see the included COPYING file */

#include "lib.h"
#include "net.h"
#include "istream.h"
#include "str.h"
#include "strescape.h"
#include "write-full.h"
#include "master-service.h"
#include "doveadm.h"
#include "doveadm-print.h"
#include "stats-settings.h"

#include <math.h>

#define DOVEADM_DUMP_DEFAULT_FIELDS \
	"count sum min max avg median stddev %95"

#define ADD_NAME_PARAM "name"
#define ADD_DESCR_PARAM "description"
#define ADD_FIELDS_PARAM "fields"
#define ADD_GROUPBY_PARAM "group-by"
#define ADD_FILTER_PARAM "filter"
#define ADD_EXPORTER_PARAM "exporter"
#define ADD_EXPORTERINCL_PARAM "exporter-include"

enum doveadm_dump_field_type {
	DOVEADM_DUMP_FIELD_TYPE_PASSTHROUGH = 0,
	DOVEADM_DUMP_FIELD_TYPE_STDDEV,
};

struct stats_cmd_context {
	string_t *cmd;
	struct doveadm_cmd_context *cctx;
	struct istream *input;
	const char *path;
	void *data;
};

struct dump_data {
	const char **fields;
	unsigned int field_count;
	enum doveadm_dump_field_type *field_types;
};

struct stats_cmd_vfuncs {
	int (*build_cmd)(struct stats_cmd_context *ctx, const char **error_r);
	void (*process_response)(struct stats_cmd_context *ctx);
};

static int build_stats_dump_cmd(struct stats_cmd_context *ctx, const char **error_r);
static int build_stats_add_cmd(struct stats_cmd_context *ctx, const char **error_r);
static int build_stats_remove_cmd(struct stats_cmd_context *ctx, const char **error_r);

static void stats_dump_process_response(struct stats_cmd_context *ctx);
static void stats_modify_process_response(struct stats_cmd_context *ctx);

static void stats_send_cmd(struct stats_cmd_context *ctx);

static struct stats_cmd_vfuncs dump_vfuncs = {
	.build_cmd = build_stats_dump_cmd,
	.process_response = stats_dump_process_response
};

static struct stats_cmd_vfuncs add_vfuncs = {
	.build_cmd = build_stats_add_cmd,
	.process_response = stats_modify_process_response
};

static struct stats_cmd_vfuncs remove_vfuncs = {
	.build_cmd = build_stats_remove_cmd,
	.process_response = stats_modify_process_response
};

static string_t *init_stats_cmd(void)
{
	string_t *cmd = t_str_new(128);
	str_append(cmd, "VERSION\tstats-reader-client\t2\t0\n");
	return cmd;
}

static void stats_exec_cmd(struct doveadm_cmd_context *cctx,
			   struct stats_cmd_vfuncs *vfuncs)
{
	struct stats_cmd_context ctx;
	const char *build_cmd_error;
	ctx.cctx = cctx;
	if (vfuncs->build_cmd(&ctx, &build_cmd_error) < 0) {
		i_error("%s", build_cmd_error);
		return;
	}
	stats_send_cmd(&ctx);
	vfuncs->process_response(&ctx);
	i_stream_destroy(&ctx.input);
}

static void handle_disconnection(struct stats_cmd_context *ctx)
{
	i_error("read(%s) failed: %s", ctx->path,
		i_stream_get_disconnect_reason(ctx->input));
}

static void stats_send_cmd(struct stats_cmd_context *ctx)
{
	int fd;
	const char *line;
	if (!doveadm_cmd_param_str(ctx->cctx, "socket-path", &ctx->path))
		ctx->path = t_strconcat(doveadm_settings->base_dir,
					"/stats-reader", NULL);

	fd = doveadm_connect(ctx->path);
	net_set_nonblock(fd, FALSE);
	if (write_full(fd, str_data(ctx->cmd), str_len(ctx->cmd)) < 0)
		i_fatal("write(%s) failed %m", ctx->path);
	ctx->input = i_stream_create_fd_autoclose(&fd, SIZE_MAX);

	if ((line = i_stream_read_next_line(ctx->input)) == NULL)
		i_fatal("%s: Failed to read VERSION line", ctx->path);
	else if (!version_string_verify(line, "stats-reader-server", 2)) {
		i_fatal_status(EX_PROTOCOL,
			"%s is not a compatible stats-reader socket", ctx->path);
	}
}

static void dump_timing(const char *const **args,
			const enum doveadm_dump_field_type field_types[],
			unsigned int fields_count)
{
	unsigned int i, args_count = str_array_length(*args);

	if (args_count > fields_count)
		args_count = fields_count;
	for (i = 0; i < args_count; i++) {
		const char *value = (*args)[i];

		switch (field_types[i]) {
		case DOVEADM_DUMP_FIELD_TYPE_PASSTHROUGH:
			break;
		case DOVEADM_DUMP_FIELD_TYPE_STDDEV: {
			double variance = strtod(value, NULL);
			value = t_strdup_printf("%.02f", sqrt(variance));
			break;
		}
		}
		doveadm_print(value);
	}
	*args += args_count;
}

static int build_stats_dump_cmd(struct stats_cmd_context *ctx,
				const char **error_r ATTR_UNUSED)
{
	bool reset;
	struct dump_data *data = t_new(struct dump_data, 1);
	const char *fields_raw;
	const char **fields;
	if (!doveadm_cmd_param_bool(ctx->cctx, "reset", &reset))
		reset = FALSE;
	if (!doveadm_cmd_param_str(ctx->cctx, "fields", &fields_raw))
		fields_raw = DOVEADM_DUMP_DEFAULT_FIELDS;

	fields = t_strsplit_spaces(fields_raw, ", ");
	data->fields = fields;
	data->field_count = str_array_length(fields);
	data->field_types =
		t_malloc0(sizeof(enum doveadm_dump_field_type) * data->field_count);
	ctx->data = data;
	ctx->cmd = init_stats_cmd();
	str_append(ctx->cmd, reset ? "DUMP-RESET" : "DUMP");
	unsigned int i;
	for (i = 0; i < data->field_count; i++) {
		str_append_c(ctx->cmd, '\t');
		if (strcmp(fields[i], "stddev") == 0) {
			data->field_types[i] = DOVEADM_DUMP_FIELD_TYPE_STDDEV;
			str_append(ctx->cmd, "variance");
		} else {
			str_append_tabescaped(ctx->cmd, fields[i]);
		}
	}
	str_append_c(ctx->cmd, '\n');
	return 0;
}

static void stats_dump_process_response(struct stats_cmd_context *ctx)
{
	unsigned int i;
	char *line;
	struct dump_data *data = ctx->data;

	doveadm_print_init(DOVEADM_PRINT_TYPE_TAB);
	doveadm_print_header_simple("metric_name");
	doveadm_print_header_simple("field");
	for (i = 0; i < data->field_count; i++)
		doveadm_print_header(data->fields[i], data->fields[i],
				     DOVEADM_PRINT_HEADER_FLAG_NUMBER);

	while ((line = i_stream_read_next_line(ctx->input)) != NULL) {
		if (line[0] == '\0')
			break;
		T_BEGIN {
			const char *const *args =
				t_strsplit_tabescaped_inplace(line);

			const char *metric_name = args[0];
			doveadm_print(metric_name); args++;
			doveadm_print("duration");
			dump_timing(&args, data->field_types, data->field_count);
			while (*args != NULL) {
				doveadm_print(metric_name);
				doveadm_print(*args); args++;
				dump_timing(&args, data->field_types, data->field_count);
			}
		} T_END;
	}
	if (line == NULL)
		handle_disconnection(ctx);
}

static int build_stats_add_cmd(struct stats_cmd_context *ctx,
			       const char **error_r)
{
	unsigned int i;
	const char *parameter;
	struct {
		const char *name;
		const char *default_val;
	} params[] = {
		{ ADD_NAME_PARAM, "" },
		{ ADD_DESCR_PARAM, "" },
		{ ADD_FIELDS_PARAM, "" },
		{ ADD_GROUPBY_PARAM, "" },
		{ ADD_FILTER_PARAM, "" },
		{ ADD_EXPORTER_PARAM, "" },
		/* Default exporter-include is to be modified
		   together with stats-settings */
		{ ADD_EXPORTERINCL_PARAM,
		  STATS_METRIC_SETTINGS_DEFAULT_EXPORTER_INCLUDE },
	};

	ctx->cmd = init_stats_cmd();
	str_append(ctx->cmd, "METRICS-ADD");

	for (i = 0; i < N_ELEMENTS(params); i++) {
		if (!doveadm_cmd_param_str(ctx->cctx, params[i].name, &parameter))
			parameter = params[i].default_val;
		if (parameter[0] == '\0' &&
		    (strcmp(params[i].name, "name") == 0 ||
		     strcmp(params[i].name, "filter") == 0)) {
			*error_r =
				t_strdup_printf("stats add: missing %s parameter",
						params[i].name);
			return -1;
		}
		str_append_c(ctx->cmd, '\t');
		str_append_tabescaped(ctx->cmd, parameter);
	}

	str_append_c(ctx->cmd, '\n');
	return 0;
}

static void stats_modify_process_response(struct stats_cmd_context *ctx)
{
	const char *line = i_stream_read_next_line(ctx->input);
	if (line == NULL) {
		handle_disconnection(ctx);
		return;
	}
	if (line[0] == '-')
		i_error("%s", ++line);
	else if (line[0] != '+')
		i_error("Invalid response: %s", line);
}

static int build_stats_remove_cmd(struct stats_cmd_context *ctx,
				  const char **error_r)
{
	const char *name;

	ctx->cmd = init_stats_cmd();
	str_append(ctx->cmd, "METRICS-REMOVE\t");

	if (!doveadm_cmd_param_str(ctx->cctx, "name", &name)) {
		*error_r = "stats remove: missing name parameter";
		return -1;
	}
	str_append_tabescaped(ctx->cmd, name);
	str_append_c(ctx->cmd, '\n');
	return 0;
}

static void doveadm_cmd_stats_dump(struct doveadm_cmd_context *cctx)
{
	stats_exec_cmd(cctx, &dump_vfuncs);
}

static void doveadm_cmd_stats_add(struct doveadm_cmd_context *cctx)
{
	stats_exec_cmd(cctx, &add_vfuncs);
}

static void doveadm_cmd_stats_remove(struct doveadm_cmd_context *cctx)
{
	stats_exec_cmd(cctx, &remove_vfuncs);
}

struct doveadm_cmd_ver2 doveadm_cmd_stats_dump_ver2 = {
	.cmd = doveadm_cmd_stats_dump,
	.name = "stats dump",
	.usage = "[-s <stats socket path>] [-r] [-f <fields>]",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('s', "socket-path", CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAM('r', "reset", CMD_PARAM_BOOL, 0)
DOVEADM_CMD_PARAM('f', "fields", CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAMS_END
};

struct doveadm_cmd_ver2 doveadm_cmd_stats_add_ver2 = {
	.cmd = doveadm_cmd_stats_add,
	.name = "stats add",
	.usage = "[--"ADD_DESCR_PARAM" <string>] "
	"[--"ADD_EXPORTER_PARAM" <name> [--"ADD_EXPORTERINCL_PARAM" <fields>]] "
	"[--"ADD_FIELDS_PARAM" <fields>] "
	"[--"ADD_GROUPBY_PARAM" <fields>] <name> <filter>",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('\0', ADD_NAME_PARAM, CMD_PARAM_STR, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAM('\0', ADD_FILTER_PARAM, CMD_PARAM_STR, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAM('\0', ADD_EXPORTER_PARAM, CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAM('\0', ADD_EXPORTERINCL_PARAM, CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAM('\0', ADD_DESCR_PARAM, CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAM('\0', ADD_FIELDS_PARAM, CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAM('\0', ADD_GROUPBY_PARAM, CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAMS_END
};

struct doveadm_cmd_ver2 doveadm_cmd_stats_remove_ver2 = {
	.cmd = doveadm_cmd_stats_remove,
	.name = "stats remove",
	.usage = "<name>",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('\0', "name", CMD_PARAM_STR, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
};
